### PR TITLE
Add batch inference

### PR DESCRIPTION
**Pull request by yhcao6**
| Model       | Batch Size | Inf Speed (imgs/s) | AP    | AP_50 | AP_75 | AP_small | AP_medium | AP_large |
|-------------|------------|--------------------|-------|-------|-------|----------|-----------|----------|
| SSD300 COCO | 1          | 22.7               | 0.257 | 0.439 | 0.262 | 0.069    | 0.277     | 0.426    |
| SSD300 COCO | 2          | 25.8               | 0.257 | 0.439 | 0.262 | 0.069    | 0.277     | 0.426    |
| SSD300 COCO | 4          | 26.1               | 0.257 | 0.439 | 0.262 | 0.069    | 0.277     | 0.426    |
| SSD300 COCO | 8          | 28.3               | 0.257 | 0.439 | 0.262 | 0.069    | 0.277     | 0.426    |

### CODE REVIEW

**Comment by stevehjc** [comment]
If `len(dataset) % samples_per_gpu != 0`, there will be an error.
Maybe the mmcv needs to be modified.
```
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> ] 5856/5863, 64.4 task/s, elapsed: 91s, ETA:     0sTraceback (most recent call last):
  File "test.py", line 300, in <module>
    main()
  File "test.py", line 260, in main
    args.gpu_collect)
  File "test.py", line 64, in multi_gpu_test
    for i, data in enumerate(data_loader):
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 582, in __next__
    return self._process_next_batch(batch)
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 608, in _process_next_batch
    raise batch.exc_type(batch.exc_msg)
AssertionError: Traceback (most recent call last):
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 99, in _worker_loop
    samples = collate_fn([dataset[i] for i in batch_indices])
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 81, in collate
    for key in batch[0]
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 81, in <dictcomp>
    for key in batch[0]
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 77, in collate
    return [collate(samples, samples_per_gpu) for samples in transposed]
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 77, in <listcomp>
    return [collate(samples, samples_per_gpu) for samples in transposed]
  File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 26, in collate
    assert len(batch) % samples_per_gpu == 0
AssertionError
```

**Comment by yhcao6** [comment]
> If `len(dataset) % samples_per_gpu != 0`, there will be an error.
> Maybe the mmcv needs to be modified.
> 
> ```
> [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> ] 5856/5863, 64.4 task/s, elapsed: 91s, ETA:     0sTraceback (most recent call last):
>   File "test.py", line 300, in <module>
>     main()
>   File "test.py", line 260, in main
>     args.gpu_collect)
>   File "test.py", line 64, in multi_gpu_test
>     for i, data in enumerate(data_loader):
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 582, in __next__
>     return self._process_next_batch(batch)
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 608, in _process_next_batch
>     raise batch.exc_type(batch.exc_msg)
> AssertionError: Traceback (most recent call last):
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 99, in _worker_loop
>     samples = collate_fn([dataset[i] for i in batch_indices])
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 81, in collate
>     for key in batch[0]
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 81, in <dictcomp>
>     for key in batch[0]
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 77, in collate
>     return [collate(samples, samples_per_gpu) for samples in transposed]
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 77, in <listcomp>
>     return [collate(samples, samples_per_gpu) for samples in transposed]
>   File "/home/han/anaconda3/envs/mmdet/lib/python3.6/site-packages/mmcv-0.2.14-py3.6-linux-x86_64.egg/mmcv/parallel/collate.py", line 26, in collate
>     assert len(batch) % samples_per_gpu == 0
> AssertionError
> ```

Just delete this assertion, I will make a Pr to mmcv.

**Comment by Jinglei5** [comment]
Thanks a lot! Could also the api implement batch inference?

**Comment by JieLi1994** [comment]
@yhcao6  thank you for your contribution，here is  a problem.  I will only get 40 results, when i test 160 images.    How to run the right result here?

File: configs/albu_example/mask_rcnn_r50_fpn_1x.py [hunk 1]
```diff
@@ -196,7 +196,7 @@
             dict(type='RandomFlip'),
             dict(type='Normalize', **img_norm_cfg),
             dict(type='Pad', size_divisor=32),
-            dict(type='ImageToTensor', keys=['img']),
+            dict(type='DefaultFormatBundle'),
```
Comment (SebastienLinker) [inline configs/albu_example/mask_rcnn_r50_fpn_1x.py]: What is the reason to modify those things? I noticed it breaks the examples using `asyncio`

Comment (JieLi1994) [reply configs/albu_example/mask_rcnn_r50_fpn_1x.py]: sry, i'm not sure

Comment (yhcao6) [reply configs/albu_example/mask_rcnn_r50_fpn_1x.py]: `DefaultFormatBundle` can convert image to [`DataContainer`](https://github.com/open-mmlab/mmcv/blob/c2c9fced2bdd5f80b254723263beca0efb60166c/mmcv/parallel/data_container.py#L19) which can pad multi images to the same shape in a batch.
This pr should update with the upstream.

**Comment by hellock** [comment]
Closed due to #3564

**Pull request closed without merging**